### PR TITLE
ASN database, part 2

### PR DESCRIPTION
**Pull request by Alphakilo**
> Conflicts needs to be resolved before merging. Closing for now, let me know if you still want this merged.

I still want it merged :smile: 
Hope you're okay with this @raunsbaekdk?

### CODE REVIEW

**Comment by raunsbaekdk** [comment]
Sure @Alphakilo!

File: README.md [hunk 1]
```diff
@@ -74,7 +79,7 @@ between IPv4 and IPv6 lookup.
 * Supports HTTPS
 * Supports common command-line clients (e.g. `curl`, `httpie`, `wget` and `fetch`)
 * JSON output
-* Country and city lookup using the MaxMind GeoIP database
+* ASN, Country and city lookup using the MaxMind GeoIP database
```
Comment (mpolden) [inline README.md]: "Country" -> "country"

File: http/http.go [hunk 2]
```diff
@@ -31,15 +31,17 @@ type Server struct {
...
+	IP                           net.IP   `json:"ip"`
+	IPDecimal                    *big.Int `json:"ip_decimal"`
+	Country                      string   `json:"country,omitempty"`
+	CountryEU                    *bool    `json:"country_eu,omitempty"`
+	CountryISO                   string   `json:"country_iso,omitempty"`
+	City                         string   `json:"city,omitempty"`
+	Hostname                     string   `json:"hostname,omitempty"`
+	Latitude                     float64  `json:"latitude,omitempty"`
+	Longitude                    float64  `json:"longitude,omitempty"`
+	AutonomousSystemNumber       string   `json:"asn,omitempty"`
```
Comment (mpolden) [inline http/http.go]: These are a bit on the long side. How about just `ASN` and `ASNOrg`?

File: http/http.go [hunk 3]
```diff
@@ -93,20 +95,27 @@ func (s *Server) newResponse(r *http.Request) (Response, error) {
...
 	country, _ := s.gr.Country(ip)
 	city, _ := s.gr.City(ip)
+	asn, _ := s.gr.ASN(ip)
 	var hostname string
 	if s.LookupAddr != nil {
 		hostname, _ = s.LookupAddr(ip)
 	}
+	var autonomousSystemNumber string
+	if asn.AutonomousSystemNumber > 0 {
+		autonomousSystemNumber = "AS" + strconv.FormatUint(uint64(asn.AutonomousSystemNumber), 10)
```
Comment (mpolden) [inline http/http.go]: Any particular reason for the uint64 cast? If not, use `strconv.Itoa`.

File: README.md [hunk 4]
```diff
@@ -37,6 +37,9 @@ EB
 
 $ curl ifconfig.co/city
 Bornyasherk
+
+$ curl ifconfig.co/asn
+AS59795,Hosting4Real
```
Comment (mpolden) [inline README.md]: I'd prefer not to have formatting in the plain text responses, how about returning just the ASN? If clients want the org string, they can use `/json`.

File: http/http.go [hunk 5]
```diff
@@ -31,15 +31,17 @@ type Server struct {
...
+	IPDecimal                    *big.Int `json:"ip_decimal"`
+	Country                      string   `json:"country,omitempty"`
+	CountryEU                    *bool    `json:"country_eu,omitempty"`
+	CountryISO                   string   `json:"country_iso,omitempty"`
+	City                         string   `json:"city,omitempty"`
+	Hostname                     string   `json:"hostname,omitempty"`
+	Latitude                     float64  `json:"latitude,omitempty"`
+	Longitude                    float64  `json:"longitude,omitempty"`
+	AutonomousSystemNumber       string   `json:"asn,omitempty"`
+	AutonomousSystemOrganization string   `json:"asn_owner,omitempty"`
```
Comment (mpolden) [inline http/http.go]: "asn_owner" -> "asn_org".

File: http/http.go (hunk 3 above)
Comment (Alphakilo) [reply http/http.go]: I'm not sure. This, like most of the code, is stolen directly from the original PR.
Let me expand: is there are particular reason why this should be an uint64? Up until that point it's an uint, as mandated by [IANA](https://tools.ietf.org/html/rfc6996).

cc @raunsbaekdk

File: README.md [hunk 6]
```diff
@@ -48,7 +51,9 @@ $ curl -H 'Accept: application/json' ifconfig.co  # or curl ifconfig.co/json
   "country": "Elbonia",
   "country_iso": "EB",
   "ip": "127.0.0.1",
-  "ip_decimal": 2130706433
+  "ip_decimal": 2130706433,
+  "asn": "AS59795",
+  "asn_owner": "Hosting4Real"
```
[the remark this answers is not in the document]
Comment (mpolden) [reply README.md]: An example should also be added to `index.html`, below "City lookup".

**Comment by Alphakilo** [comment]
Should be good to go.  Thanks for you patience,  @mpolden :slightly_smiling_face:

**Comment by mpolden** [comment]
There's a conflict that must be resolved. Can you also rebase all commits into one?

**Comment by maximbaz** [comment]
Github allows you as a project owner to rebase all commits into, see the button "Squash and merge" 😉

**Comment by Alphakilo** [comment]
Best I could do...

**Comment by maximbaz** [comment]
Would it make sense to just add _all_ GeoLite2 databases? Besides of what we currently use, there also are:

- Anonymous IP Database (e.g. detects VPN usage)
- Connection Type Database (Cable / Cellular / etc)
- Domain Database (reverse DNS basically)
- ISP Database

**Comment by raunsbaekdk** [comment]
@mpolden 

May I suggest that you fixes the domain part:
https://github.com/mpolden/echoip/blob/7fbc2e1b9fc80146e5fa16bde1ab636ef63725fa/index.html#L84
To: $ http {{ .Host }}/asn

**Comment by patschi** [comment]
I've opened #68 for that - it's just a simple fix and needs to be merged.

**Comment by mpolden** [comment]
Fixed.

**Pull request merged**